tests/queries/0_stateless/03001_memory_table_circ_buffer_usage.sql

```diff
@@ -0,0 +1,41 @@
+DROP TABLE IF EXISTS memory;
```

Comment:
interesting to see if `min_bytes_to_keep` will be bigger than `max_bytes_to_keep` (the same with rows)

Reply:
That is a great thought. If that was the case, then I can imagine this would lead to a reduction in the number of `Block` deletions as we would be breaking the `while` loop more frequently.

I was tempted to put a check in to ensure the invariant holds. I thought I would give the user the benefit of the doubt but, in hindsight, it's probably better to be safe rather than sorry. I'll add a check in to verify our upper and lower bounds make sense before acting on them. :)

Reply:
Actually, that also raises the question. Should we always check if these values are provided as pairs? (I would also update the documentation to reflect this too.)

Reply:
I've updated the while loop to check for correct upper and lower bounds. I've also updated `SettingsChangeHistory` but I need to comb through the rest of the tests and the seg-fault tomorrow (need to go to bed and let this thing compile for now). Thanks for the help Yarik, I'll keep you updated! 🌞

Reply:
> Actually, that also raises the question. Should we always check if these values are provided as pairs? (I would also update the documentation to reflect this too.)

I think that generally users will use both of them at the same time, but in some cases users can use `max_bytes_to_keep` only. So, I think it logical to allow people to use `max_bytes_to_keep` without `min_bytes_to_keep`, but not `min_bytes_to_keep ` without `max_bytes_to_keep`(it doesn't make sense)🙂

Reply:
This sounds good. I will also update the docs to specify the rules that `max_...` is allowed on it's own but `min_...` must be accompanied by `max_...` setting.

If invalid values are given, I will throw an exception I imagine instead of failing silently.